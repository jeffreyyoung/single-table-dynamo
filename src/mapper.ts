import { takeWhile } from './utils/takeWhile';

export type KeysOfType<T, TProp> = { [P in keyof T]: T[P] extends TProp? P : never }[keyof T];

export enum IndexType {
  Primary = 'Primary',
  Secondary = 'Secondary'
}

export type PrimaryIndex<ID, IndexTagNames = string> = Index<IndexTagNames> | (CompositeIndex<ID, IndexTagNames>)

export type SecondaryIndex<Src, IndexTagNames = string> = NamedIndex<IndexTagNames> | (NamedIndex<IndexTagNames> & CompositeIndex<Src, IndexTagNames>);

export function isSecondaryIndex<Src, IndexTagNames>(i: Index<IndexTagNames>): i is SecondaryIndex<Src, IndexTagNames> {
  let namedIndex = i as SecondaryIndex<Src>;
  if (namedIndex.indexName) {
    return true;
  }

  return false;
}

type NamedIndex<IndexTagNames = string> = Index<IndexTagNames> & { indexName: string };

export type Index<IndexTagNames = string> = {
  tag?: IndexTagNames
  partitionKey: string
  sortKey?: string
};

type CompositeKeyField<Src> = (KeysOfType<Src, string> | NonStringField<Src>);

export type CompositeIndex<Src, IndexTagNames = string> = {
  fields: CompositeKeyField<Src>[]
} & Index<IndexTagNames>;

interface NonStringField<Src> {
  toString: (s: Src) => string
  fields: Extract<keyof Src, string>[]
}


export type MapperArgs<Id, Src, IndexTagNames = string> = {
  typeName: string,
  indexFieldSeparator?: string,
  primaryIndex: PrimaryIndex<Id, IndexTagNames>,
  secondaryIndexes?: SecondaryIndex<Src, IndexTagNames>[]
}



function isNonStringField<T>(thing: any): thing is NonStringField<T> {
  return thing?.fields && thing?.toString;
}

/**
 * The mapper is what has the responsibility of decorating
 * a 
 * @param args 
 */
export class Mapper<Id, Src, IndexTagNames = string> {
  readonly args: MapperArgs<Id, Src, IndexTagNames>

  constructor(args: MapperArgs<Id, Src, IndexTagNames>) {
    this.args = args;
  }

  getKey(id: Id) {
    return this.computeIndexFields(id, this.args.primaryIndex);
  }

  indexes() {
    return [this.args.primaryIndex, ...(this.args.secondaryIndexes || [])];
  }
  /**
   * Takes an object to be saved to ddb,
   * and adds any computed index fields to it
   * @param src 
   */
  decorateWithIndexedFields(src: Src) {
    const res = {...src};

    this.indexes().forEach((i: Index<IndexTagNames>) => {
      Object.assign(res, this.computeIndexFields(src, i));
    });
    return res;
  }

  stringifyField(src: Partial<Src>, f: CompositeKeyField<Src>) {
    if (isNonStringField(f)) {
      return f.toString(src);
    } else {
      return src[f as string];
    }
  }

  _isCompositeIndex(index: Index<IndexTagNames>): index is CompositeIndex<Src, IndexTagNames> {
    return 'fields' in index;
  }

  _computeCompositePrimaryKey(src: Partial<Src>, primaryField: CompositeIndex<Src>['fields'][number]) {
    const field = this.stringifyField(src, primaryField);
    if (!field) {
      throw new Error(`You must provide partition key fields, required field: ${primaryField}, provided fields: ${JSON.stringify(src, null, 1)}`)
    }
    return [this.args.typeName, this.stringifyField(src, primaryField)].join('#');
  }

  _computeCompositeSortKey(src: Partial<Src>, sortKeyFields: CompositeIndex<Src>['fields']) {
    return [this.args.typeName, ...sortKeyFields.map(f => this.stringifyField(src, f))].join('#')
  }

  _isInSrc(src: Partial<Src>, indexField: CompositeIndex<Src>['fields'][number]) {
    let fields = [];
    if (isNonStringField(indexField)) {
      fields = [...indexField.fields]
    } else {
      fields = [indexField]
    };
    return fields.every(f=> (src as any).hasOwnProperty(f))
  }

  /**
   * Returns the key/values of an index
   * @param src 
   * @param typeName 
   * @param index 
   */
  computeIndexFields(src: Partial<Src>, index: Index<IndexTagNames>) {
    if (this._isCompositeIndex(index)) {
      if (index.fields.length < 1) {
        throw new Error(`A partition key field must be provided:\n\nprovided fields: ${JSON.stringify(src,null,1)}\nindex: ${JSON.stringify(index,null,1)}`)
      }

      const partitionValue = this._computeCompositePrimaryKey(src, index.fields[0]);

      const sortValue = this._computeCompositeSortKey(src, takeWhile(
          index.fields.slice(1),
          (f) => !this._isInSrc(src, f)
        ));

      return {
        ...partitionValue && {[index.partitionKey]: partitionValue},
        ...sortValue && index.sortKey && {[index.sortKey]: sortValue}
      }
    } else {
      return {
        [index.partitionKey]: src[index.partitionKey],
        ...index.sortKey && src[index.sortKey] && {[index.sortKey]: src[index.sortKey]}
      }
    }
  }
}
