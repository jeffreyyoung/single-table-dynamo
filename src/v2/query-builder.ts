type Operator = "EQ" | "NE" | "IN" | "LE" | "LT" | "GE" | "GT" | "BETWEEN" | "NOT_NULL" | "NULL" | "CONTAINS" | "NOT_CONTAINS" | "BEGINS_WITH"
type Where = {
  fieldName: string
  operator: Operator
  value: string | number
}

/**
 * Encapsulates all data needed to generate
 * query
 */
export type QueryData = {
  keyConditions: Where[]
  sortOrder: 'asc' | 'desc'
  limit: number
  indexName?: string
  tableName?: string
  /**
   * The key of the last fetched object
   */
  cursor?: object
}

export class QueryBuilder {
  data: QueryData
  constructor() {
    this.data = {
      keyConditions: [],
      sortOrder: 'desc',
      limit: 25
    }
  }

  table(tabeName: string) {
    this.data.tableName = tabeName;
    return this;
  }

  index(indexName: string) {
    this.data.indexName = indexName;
    return this;
  }

  sort(direction: 'asc' | 'desc') {
    this.data.sortOrder = direction;
    return this;
  }

  limit(l: number) {
    this.data.limit = l;
    return this;
  }

  cursor(l: object) {
    this.data.cursor = l;
    return this;
  }

  where(key: string, op: Operator, value: string | number) {
    this.data.keyConditions.push({
      fieldName: key,
      operator: op,
      value
    });
    return this;
  }

  build() {
    return {
      TableName: this.data.tableName,
      ScanIndexForeward: this.data.sortOrder === 'asc' ? true : false,
      Select: 'ALL_ATTRIBUTES',
      ...this.data.indexName && {IndexName: this.data.indexName},
      ...this._buildConditionExpression(),
      ...this._buildCursor()
    }
  }

  _buildConditionExpression() {
    const ExpressionAttributeNames = {};
    const KeyConditionExpression: string[] = [];
    const ExpressionAttributeValues = {};

    this.data.keyConditions.forEach((condition, i) => {
      const attributeName = 'attr'+i;
      ExpressionAttributeNames[`#${attributeName}`] = condition.fieldName;
      ExpressionAttributeValues[`:${attributeName}`] = condition.value;
      KeyConditionExpression.push(`#${attributeName} ${condition.operator} :${attributeName}`)
    })
    return {
      ExpressionAttributeNames,
      ExpressionAttributeValues,
      KeyConditionExpression: KeyConditionExpression.join(' and ')
    }
  }

  _buildCursor() {
    if (this.data.cursor) {
      return {
        ExclusiveStartKey: this.data.cursor
      }
    }
    return {}
  }
}