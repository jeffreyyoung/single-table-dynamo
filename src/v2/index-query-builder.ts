import { isPrimaryIndex, Mapper, SingleTableIndex } from "./mapper";
import { QueryBuilder } from './query-builder';

export class IndexQueryBuilder<Src> {
  mapper: Mapper<Src>
  builder: QueryBuilder
  index: SingleTableIndex<Src>

  constructor(tableName: string, index: SingleTableIndex<Src>, mapper: Mapper<Src>) {
    this.mapper = mapper;
    this.index = index;
    this.builder = new QueryBuilder();
    this.builder
      .table(tableName)
    if (!isPrimaryIndex(index)) {
      this.builder.index(index.indexName);
    }
  }

  limit(t: number) {
    this.builder.limit(t);
    return this;
  }

  sort(direction: 'asc' | 'desc') {
    this.builder.sort(direction);
    return this;
  }

  cursor(str: 'string') {
    this.builder.cursor(JSON.parse(str));
    return this;
  }

  build() {
    return this.builder.build();
  }

  where(src: Partial<Src>) {
    const indexes = this.mapper.computeIndexFields(src, this.index) as any;
    if (indexes[this.index.partitionKey]) {
      this.builder.where(this.index.partitionKey as any, 'EQ', indexes[this.index.partitionKey])
    }

    if (indexes[this.index.sortKey]) {
      this.builder.where(this.index.sortKey as any, 'BEGINS_WITH', indexes[this.index.sortKey])
    }

    return this;
  }
}