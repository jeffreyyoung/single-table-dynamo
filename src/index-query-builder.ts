import { DocumentClient } from "aws-sdk/clients/dynamodb";
import {
  Mapper,
  ifSecondaryIndexGetName,
  IndexBase,
  RawResult,
} from "./mapper";
import { QueryBuilder } from "./query-builder";
import { AnyRepository } from "./repository";

export function decodeCursor(cursor: string): object {
  const decodedCursor = Buffer.from(cursor, "base64").toString("utf-8");
  return JSON.parse(decodedCursor);
}

export function getCursorEncoder<Src extends object>(args: {
  primaryIndex: IndexBase<Src>;
  secondaryIndex: IndexBase<Src>;
  mapper: Mapper;
}) {
  return (src: Src) => {
    const json = JSON.stringify({
      ...args.mapper.getIndexKey(src, args.primaryIndex),
      ...(args.secondaryIndex &&
        args.mapper.getIndexKey(src, args.secondaryIndex)),
    });

    return Buffer.from(json).toString("base64");
  };
}

type IndexQueryBuilderArgs<T> = {
  tableName: string;
  mapper: Mapper;
  parseAndMigrate: AnyRepository["parseAndMigrate"];
  index: IndexBase<T>;
  builder?: QueryBuilder;
  ddb: DocumentClient;
};

export class IndexQueryBuilder<Src extends object> {
  tableName: string;
  mapper: AnyRepository["mapper"];
  index: IndexBase<Src>;
  parseAndMigrate: AnyRepository["parseAndMigrate"];
  builder: QueryBuilder;
  ddb: DocumentClient;
  encodeCursor: (src: Src) => string;
  decodeCursor = decodeCursor;

  constructor(args: IndexQueryBuilderArgs<Src>) {
    this.tableName = args.tableName;
    this.mapper = args.mapper;
    this.index = args.index;
    this.ddb = args.ddb;
    this.parseAndMigrate = args.parseAndMigrate;
    this.builder = (args.builder || new QueryBuilder()).table(args.tableName);

    this.encodeCursor = getCursorEncoder<any>({
      secondaryIndex: args.index,
      primaryIndex: args.mapper.args.primaryIndex,
      mapper: args.mapper,
    });
    const secondaryIndexName = ifSecondaryIndexGetName(args.index);
    if (secondaryIndexName) {
      this.builder = this.builder.index(secondaryIndexName);
    }
  }

  clone(builder: QueryBuilder = this.builder) {
    return new IndexQueryBuilder<Src>({
      tableName: this.tableName,
      mapper: this.mapper,
      index: this.index,
      ddb: this.ddb,
      parseAndMigrate: this.parseAndMigrate,
      builder: builder.cloneWith(),
    });
  }

  limit(t: number) {
    return this.clone(this.builder.limit(t));
  }

  sort(direction: "asc" | "desc") {
    return this.clone(this.builder.sort(direction));
  }

  cursor(str: string) {
    const decoded = this.decodeCursor(str);
    return this.clone(this.builder.cursor(decoded));
  }

  build() {
    return this.builder.build();
  }

  async exec() {
    const expression = this.builder.build();
    this.mapper.args.on?.queryStart?.(expression);
    const _res = await this.ddb.query(expression as any).promise();
    const res = {
      ..._res,
      Items: (await Promise.all(
        (_res.Items || []).map((item) => {
          return this.parseAndMigrate(item);
        })
      )) as Src[],
    };
    const hookInfo: RawResult[] =
      _res.Items?.flatMap((Item) =>
        Item
          ? [this.mapper.getHookResultInfo(this.mapper.parseId(Item), Item)]
          : []
      ) ?? [];
    if (this.mapper.args.dataLoader) {
      for (const result of hookInfo) {
        this.mapper.args.dataLoader.clear(result).prime(result, {
          Item: result.Item || undefined,
          $response: {} as any,
        });
      }
    }
    this.mapper.args.on?.query?.(expression, hookInfo);

    return Object.assign(res, {
      encodeCursor: this.encodeCursor,
      lastCursor: res.Items?.length
        ? this.encodeCursor(res.Items[res.Items.length - 1])
        : undefined,
    });
  }

  /**
   * Executes the query and returns the first item in the result set.
   * @returns The first item in the result set.
   */
  async execOne(): Promise<Src | undefined> {
    const res = await this.limit(1).exec();
    return res.Items?.[0];
  }

  /**
   * Iterater over the query result set in batches
   *
   *
   *
   * @returns An async iterator that yields the result set in batches.
   * @example
   *    for await (const items of queryBuilder.execAll()) {
   *      console.log(items);
   *    }
   */
  async *execAll(): AsyncGenerator<Src[]> {
    let cursor: string | null = null;
    type Page = {
      Items: Src[];
      lastCursor?: string;
    };
    do {
      const page: Page = await (cursor
        ? this.cursor(cursor).exec()
        : this.exec());
      if (page.Items?.length) {
        yield page.Items;
      }
      cursor = page.lastCursor || null;
    } while (cursor);
  }

  where(src: Partial<Src>) {
    let builder = this.builder;
    const indexedKeyToValue = this.mapper.getIndexKey(src as Src, this.index, {
      partial: true,
    }) as any;
    const hasEveryField = this.index.fields.every((f) => hasOwn(src, f));
    if (indexedKeyToValue[this.index.pk]) {
      builder = builder.where(
        this.index.pk as any,
        "=",
        indexedKeyToValue[this.index.pk]
      );
    }

    if (indexedKeyToValue[this.index.sk!]) {
      if (hasEveryField) {
        builder = builder.where(
          this.index.sk as any,
          "=",
          indexedKeyToValue[this.index.sk!]
        );
      } else {
        builder = builder.where(
          this.index.sk as any,
          "BEGINS_WITH",
          indexedKeyToValue[this.index.sk!] // TODO + '#'
        );
      }
    }

    return this.clone(builder);
  }
}

function hasOwn(thing: object, field: string) {
  if (typeof thing === "object") {
    return Object.prototype.hasOwnProperty.call(thing, field);
  }
  return false;
}
