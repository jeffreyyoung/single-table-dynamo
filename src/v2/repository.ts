import { IndexQueryBuilder } from './index-query-builder'
import { MapperArgs, Mapper } from './mapper'

type RepoArgs<Src> = {
  tableName: string
} & MapperArgs<Src>


export  class Repository<ID, Src> {
  args: RepoArgs<Src>
  mapper: Mapper<Src>

  constructor(args: RepoArgs<Src>) {
    this.args = args;
    this.mapper = new Mapper<Src>(args);
  }
  get(id: ID){

  }
  put(src: Src){}
  update(src: Partial<Src> & ID){}
  delete(id: Src){}
  query(tag: string) {
    const builder = new IndexQueryBuilder(
      this.args.tableName,
      this.indexByTag(tag),
      this.mapper
    );
    return builder
  }
  indexByTag(tag: string) {
    const index = this.args.indexes.find(i => i.tag === tag)
    if (!index) {
      throw new Error(`No index exists for that tag, tag: ${tag}, args: ${JSON.stringify(this.args, null, 3)}`)
    }
    return index;
  }
}