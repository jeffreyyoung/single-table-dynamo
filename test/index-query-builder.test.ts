import { IndexQueryBuilder } from './../src/v2/index-query-builder';
import { Mapper } from './../src/v2/mapper';

type User = {
  id: string;
  state: string;
  country: string;
  createdAt: string;
  updatedAt: string;
  count: number;
};

// const indexes = [
//   {
//     partitionKey: 'pk1',
//     sortKey: 'sk1',
//   },
//   {
//     indexName: 'third',
//     partitionKey: 'pk2',
//     sortKey: 'sk2',
//   },
// ];

const mapper = new Mapper<User>({
  typeName: 'User',
  indexes: [
    {
      partitionKey: 'pk1',
      sortKey: 'sk1',
      tag: 'countryByStateByCreatedAt',
      fields: ['country', 'state', 'createdAt'],
    },
    {
      indexName: 'third',
      partitionKey: 'pk2',
      sortKey: 'sk2',
      tag: 'stateByCountryByYeehaw',
      fields: [
        'state',
        'country',
        { toString: () => 'yeehaw', fields: ['count'] },
      ],
    },
    {
      indexName: 'countryByUpdatedAt',
      tag: 'meowowow',
      partitionKey: 'state',
      sortKey: 'country',
    },
  ],
});

const getBuilder = (which: number) =>
  new IndexQueryBuilder<User>('yeehaw', mapper.args.indexes[which], mapper);

test('should build query with no sortkey', () => {
  expect(
    getBuilder(0)
      .where({
        country: 'USA',
      })
      .limit(25)
      .build()
  ).toEqual({
    ExpressionAttributeNames: {
      '#attr0': 'pk1',
    },
    ExpressionAttributeValues: {
      ':attr0': 'User#USA',
    },
    KeyConditionExpression: '#attr0 EQ :attr0',
    ScanIndexForeward: false,
    Select: 'ALL_ATTRIBUTES',
    TableName: 'yeehaw',
  });
});

test('should build query with extra fields', () => {
  expect(
    getBuilder(0)
      .where({
        country: 'USA',
        createdAt: '2010-10-21',
      })
      .limit(25)
      .build()
  ).toEqual({
    ExpressionAttributeNames: {
      '#attr0': 'pk1',
    },
    ExpressionAttributeValues: {
      ':attr0': 'User#USA',
    },
    KeyConditionExpression: '#attr0 EQ :attr0',
    ScanIndexForeward: false,
    Select: 'ALL_ATTRIBUTES',
    TableName: 'yeehaw',
  });
});

test('should build query with sortkey', () => {
  expect(
    getBuilder(0)
      .where({
        country: 'USA',
        state: 'UT',
      })
      .limit(25)
      .build()
  ).toEqual({
    ExpressionAttributeNames: {
      '#attr0': 'pk1',
      '#attr1': 'sk1',
    },
    ExpressionAttributeValues: {
      ':attr0': 'User#USA',
      ':attr1': 'UT',
    },
    KeyConditionExpression: '#attr0 EQ :attr0 and #attr1 BEGINS_WITH :attr1',
    ScanIndexForeward: false,
    Select: 'ALL_ATTRIBUTES',
    TableName: 'yeehaw',
  });
});

test('should throw with no partition key', () => {
  expect(() =>
    getBuilder(0)
      .where({
        state: 'UT',
      })
      .limit(25)
      .build()
  ).toThrowErrorMatchingSnapshot();
});

test('should build non primary index', () => {
  expect(
    getBuilder(1)
      .where({
        state: 'WA',
      })
      .build()
  ).toEqual({
    ExpressionAttributeNames: {
      '#attr0': 'pk2',
    },
    ExpressionAttributeValues: {
      ':attr0': 'User#WA',
    },
    KeyConditionExpression: '#attr0 EQ :attr0',
    IndexName: "third",
    ScanIndexForeward: false,
    Select: 'ALL_ATTRIBUTES',
    TableName: 'yeehaw',
  });
});
