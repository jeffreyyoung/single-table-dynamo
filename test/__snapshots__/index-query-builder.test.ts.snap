// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw with no partition key 1`] = `
"You must provide partition key fields, required field: country, provided fields: {
 \\"state\\": \\"UT\\"
}"
`;
